dali/operators/nvcvop/nvcvop.cc

```diff
@@ -0,0 +1,141 @@
+    case DALI_UINT16:
+    case DALI_UINT32:
+    case DALI_UINT64:
+      return nvcv::DataKind::UNSIGNED;
+    case DALI_INT8:
+    case DALI_INT16:
+    case DALI_INT32:
+    case DALI_INT64:
+      return nvcv::DataKind::SIGNED;
+    case DALI_FLOAT16:
+    case DALI_FLOAT:
+    case DALI_FLOAT64:
+      return nvcv::DataKind::FLOAT;
+    default:
+      DALI_FAIL("Unsupported data type");
+  }
+}
+
+nvcv::Swizzle GetSimpleSwizzle(int num_channels) {
+  if (num_channels == 1) {
+      return nvcv::Swizzle::S_X000;
+  } else if (num_channels == 2) {
+      return nvcv::Swizzle::S_XY00;
+  } else if (num_channels == 3) {
+      return nvcv::Swizzle::S_XYZ0;
+  } else if (num_channels == 4) {
+      return nvcv::Swizzle::S_XYZW;
+  } else {
+      DALI_FAIL("Unsupported number of channels");
+  }
+}
+
+nvcv::ImageFormat GetImageFormat(DALIDataType dtype, int num_channels) {
+  nvcv::MemLayout mem_layout = nvcv::MemLayout::PITCH_LINEAR;
+  nvcv::DataKind data_kind = GetDataKind(dtype);
+  nvcv::Swizzle swizzle = GetSimpleSwizzle(num_channels);
+  nvcv::PackingParams packing_params{};
+  packing_params.byteOrder = nvcv::ByteOrder::MSB;
+  packing_params.swizzle = swizzle;
+  for (int c = 0; c < num_channels; ++c) {
+    packing_params.bits[c] = TypeTable::GetTypeInfo(dtype).size() * 8;
+  }
+
+  auto packing = nvcv::MakePacking(packing_params);
+  return nvcv::ImageFormat(mem_layout, data_kind, swizzle, packing);
+}
+
+/**
+ * @brief Wrap tensor data buffer into nvcv::Image.
+ *
+ * @param data pointer to beginning of the buffer
+ * @param shape underlying tensor shape
+ * @param format image format
+ */
+nvcv::Image WrapImage(void *data, const TensorShape<> &shape, const nvcv::ImageFormat &format) {
+  DALI_ENFORCE(format.numPlanes() == 1);
+  auto pixel_size = format.planeBitsPerPixel(0) / 8;
+  nvcv::ImageDataStridedCuda::Buffer buf{};
+  buf.numPlanes = 1;
+  buf.planes[0].basePtr = static_cast<NVCVByte*>(data);
+  buf.planes[0].height = shape[0];
+  buf.planes[0].width  = shape[1];
+  buf.planes[0].rowStride = shape[1] * pixel_size;
+  nvcv::ImageDataStridedCuda img_data(format, buf);
+  return nvcv::ImageWrapData(img_data);
+}
+
+nvcv::Image AsImage(SampleView<GPUBackend> sample, const nvcv::ImageFormat &format) {
+  return AsImage(ConstSampleView<GPUBackend>(sample), format);
+}
+
+nvcv::Image AsImage(ConstSampleView<GPUBackend> sample, const nvcv::ImageFormat &format) {
+  auto &shape = sample.shape();
+  auto data = const_cast<void*>(sample.raw_data());
+  return WrapImage(data, shape, format);
+}
+
+void AllocateImagesLike(const TensorList<GPUBackend> &t_list,
+                        kernels::DynamicScratchpad &scratchpad, nvcv::ImageBatchVarShape &output) {
+  auto channel_dim = t_list.GetLayout().find('C');
+  uint8_t *buffer = scratchpad.AllocateGPU<uint8_t>(t_list.nbytes());
+  size_t offset = 0;
+  for (int s = 0; s < t_list.num_samples(); ++s) {
+    auto num_channels = (channel_dim >= 0) ? t_list[s].shape()[channel_dim] : 1;
```

Comment:
1. TensorList's `operator[]` is complex enough to warrant having a temporary local variable instead of re-invoking it several times below.
2. I think that it'd be nice (and saving us trouble in the future) if we added an option to get a `shared_ptr` for a sample in TensorList. The SampleView, which was intended only for data access (as opposed to data ownership) doesn't provide that. With a shared_ptr we could provide a custom data deleter to the resulting NVCV tensor, thus making it safe for the NVCV object to outlive the DALI tensor.

Reply:
1. Done
2. I added the info about lifetime bounds to the docs